### DISCUSSION TITLE
FV elemental vs qp evaluation

### DISCUSSION THREAD

**Discussion by heinono1** [idaholab/moose#25199] [Q&A General] [answered]
Hi. I think I have my FV version of Ferret (module for ferroic systems) working. Some of the FV kernels are elemental, some are flux kernels. It seems to me that when evaluating the residuals for elemental kernels, I can either evaluate variables/functors on quadrature points and would as an example have the residual as
```
    return prefactor * (_mag_y[_qp]* _nz[_qp] - _mag_z[_qp]* _ny[_qp]);
```
where _mag_x, _nx etc are functors and prefactor is ADReal,
or I can define elemental values like
```
  ADReal _mag_x_elem = _mag_x(makeElemArg(_current_elem),determineState());
  ADReal _mag_y_elem = _mag_y(makeElemArg(_current_elem),determineState());
```
and similarly for _ny_elem and _ny_elem and evaluate the residual as
```
 return  prefactor * (_mag_z_elem * _nx_elem - _mag_x_elem * _nz_elem);
```
Am I correct that either one works, and, if so, is there any particular advantage to one vs the other?

**Comment by GiudGiud** [comment, marked as answer]
Hello

Both formulations are equivalent in terms of results for a finite volume variable.
The element argument on a constant monomial gets the cell-center value, which is also the only quadrature point in a FV simulation (_qp = 0).

The first formulation would be a tad faster (possibly not measurable), as it's a more lightweight interface
The second formulation (using the functor routines) will let you use functions, variables, functor material properties and soon postprocessors for mag_x and _y.

@lindsayad 

As you prefer basically!
Guillaume

**Reply by heinono1** [reply]
Thanks, Guillaume. That is what I was guessing. Incidentally, I did write postprocessors for the FV code but had to use ADFunctorElementalAux and ADFunctorElementalGradientAux to define elemental variables and their gradients that I could use with the ElementIntegralPostprocessor, so postprocessors that can use the FV variables directly will be great!

**Reply by GiudGiud** [reply]
It's coming real soon! Josh has fixed a legacy issues with postprocessors and now we are on the brink of having that

**Reply by heinono1** [reply]
That will be great!

**Reply by lindsayad** [reply]
I will say that I think it's easier to support finite volume with the functor interfaces, e.g. `makeElemArg`. So we may develop FV features in the future that are only supported through the functor interfaces. But the `_qp` code that already exists will probably always be supported, as there is substantial user code that relies on it.
